source/extensions/access_loggers/open_telemetry/grpc_access_log_impl.cc

```diff
@@ -43,15 +43,24 @@ opentelemetry::proto::common::v1::KeyValue getStringKeyValue(const std::string&
 } // namespace
 
 // See comment about the structure of repeated fields in the header file.
-void GrpcAccessLoggerImpl::initMessageRoot(const std::string& log_name,
-                                           const LocalInfo::LocalInfo& local_info) {
+void GrpcAccessLoggerImpl::initMessageRoot(
+    const envoy::extensions::access_loggers::grpc::v3::CommonGrpcAccessLogConfig& config,
+    const LocalInfo::LocalInfo& local_info) {
   auto* resource_logs = message_.add_resource_logs();
   root_ = resource_logs->add_instrumentation_library_logs();
   auto* resource = resource_logs->mutable_resource();
-  *resource->add_attributes() = getStringKeyValue("log_name", log_name);
+  *resource->add_attributes() = getStringKeyValue("log_name", config.log_name());
   *resource->add_attributes() = getStringKeyValue("zone_name", local_info.zoneName());
   *resource->add_attributes() = getStringKeyValue("cluster_name", local_info.clusterName());
   *resource->add_attributes() = getStringKeyValue("node_name", local_info.nodeName());
+
+  // TODO: support other type
+  for (const auto& custom_tag : config.custom_tags()) {
```

Comment:
I think it would be better to use Envoy::Tracing::CustomTag and Envoy::Tracing::CustomTagUtility here as it will offer you the generic solution. The problem here is that the CustomTag::applyLog accepts specific message type and does not support OT message. A possible solution is to add the `virtual absl::string_view value(const CustomTagContext&) const PURE;` method to the Envoy::Tracing::CustomTag interface. When you should be able to get both key and value for each custom tag and store them as attrbutes in the OT access log message.

What do you think?

Reply:
I spent a lot of time to use `CustomTagUtility` here, but it would made a lot of changes. I'm not that familiar with all of the related codes, also only literal and env type would be able to use here(there's no request context).

@yanavlasov if another PR add those changes to `CustomTagUtility`, I'm ok to refactor my implements.

Reply:
Ok, I understand. The [custom_tag](https://github.com/envoyproxy/envoy/blob/4afdf50dfccb3af029a8477a322341e5d42b039c/api/envoy/extensions/access_loggers/grpc/v3/als.proto#L103) was originally intended to per per log message. OT did not use it and introduced the more flexible mechanism using the [attributes](https://github.com/envoyproxy/envoy/blob/4afdf50dfccb3af029a8477a322341e5d42b039c/api/envoy/extensions/access_loggers/open_telemetry/v3/logs_service.proto#L41) formatters. And no you'd like to reuse the original custom_tags mechanism for initializing attributes of log request root. I need to think about this. I'll comment on this tomorrow.

Reply:
@yanavlasov any update?

Reply:
Sorry for the delay. My thinking is that it would be confusing to reuse `custom_tags` from common config proto as they are applied to each access log message. And you are applying these attributes to the batch of access log messages. I think a better approach would be to add a new field to the [OpenTelemetryAccessLogConfig](https://github.com/envoyproxy/envoy/blob/5e9ca8d520873ae6a8607a0585ac4466b2d2d19a/api/envoy/extensions/access_loggers/open_telemetry/v3/logs_service.proto#L42), similar to the existing `attributes` field. You will need to modify the [GrpcAccessLoggerCacheImpl::createLogger()](https://github.com/envoyproxy/envoy/blob/5e9ca8d520873ae6a8607a0585ac4466b2d2d19a/source/extensions/access_loggers/open_telemetry/grpc_access_log_impl.cc#L74) method to accept the OpenTelemetryAccessLogConfig and not the CommonGrpcAccessLogConfig. 

I think it could be cleaner. What do you think?

Reply:
+1

Reply:
> Sorry for the delay. My thinking is that it would be confusing to reuse `custom_tags` from common config proto as they are applied to each access log message. And you are applying these attributes to the batch of access log messages. I think a better approach would be to add a new field to the [OpenTelemetryAccessLogConfig](https://github.com/envoyproxy/envoy/blob/5e9ca8d520873ae6a8607a0585ac4466b2d2d19a/api/envoy/extensions/access_loggers/open_telemetry/v3/logs_service.proto#L42), similar to the existing `attributes` field. You will need to modify the [GrpcAccessLoggerCacheImpl::createLogger()](https://github.com/envoyproxy/envoy/blob/5e9ca8d520873ae6a8607a0585ac4466b2d2d19a/source/extensions/access_loggers/open_telemetry/grpc_access_log_impl.cc#L74) method to accept the OpenTelemetryAccessLogConfig and not the CommonGrpcAccessLogConfig.
> 
> I think it could be cleaner. What do you think?

Look better than now, one more concern is what should we deal with the custom_tags in CommonGrpcAccessLogConfig, just leave them there?

Reply:
Yes, we'll just leave them there. They are used in Envoy's "native" gRPC access logger.